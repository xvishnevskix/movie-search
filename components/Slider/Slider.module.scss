.container {
    width: 100%;
}

.inputs {
    display: flex;
    align-items: flex-end;
    margin-bottom: 25px;
}

.input {
    width: 100%;
    
    input {
        width: 100%;
    }

    &:not(:last-child) {
        margin-right: 10px;
    }
}