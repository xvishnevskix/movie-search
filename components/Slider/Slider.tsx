import { ActionCreatorWithPayload } from "@reduxjs/toolkit";
import { ChangeEvent, FC, useState } from "react"
import { Input } from "@/components/Input/Input";
import Nouislider from "nouislider-react"
import styles from './Slider.module.scss'

interface SliderProps {
  min: number;
  max: number;
  step?: number;
  startMin: number;
  startMax: number;
  setMin: ActionCreatorWithPayload<number>;
  setMax: ActionCreatorWithPayload<number>;
  setPage: ActionCreatorWithPayload<number>;
}

export const Slider: FC<SliderProps> = ({ min, max, startMin, startMax, step, setMin, setMax, setPage}) => {

  const [inputHandle, setInputHandle] = useState({left: startMin, right: startMax});
  const leftInputHandle = Math.ceil(Number(inputHandle.left))
  const rightInputHandle = Math.ceil(Number(inputHandle.right))

  const handleSlider = (sliderVal: number[]) => {

    setInputHandle({
      left: sliderVal[0],
      right: sliderVal[1]
    })
  }

  const handleSliderChange = () => {
    setMin(leftInputHandle)
    setMax(rightInputHandle)
    setPage(1)
  }

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputHandle({...inputHandle, [e.target.name]: e.target.value})
    handleSliderChange()
  }

  return (
    <div className={styles.container}>
      <div className={styles.inputs}>
        <Input
          classN={styles.input}
          name="left"
          type="number"
          placeholder={min?.toString()}
          value={leftInputHandle}
          onChange={handleChange}
          min={min}
          max={max}
        />
        <Input
          classN={styles.input}
          name="right"
          type="number"
          placeholder={max?.toString()}
          value={rightInputHandle}
          onChange={handleChange}
          min={min}
          max={max}
        />
      </div>
      <Nouislider
        onUpdate={handleSlider}
        onChange={handleSliderChange}
        range={{ min: min, max: max }}
        animate={false}
        step={step}
        start={[inputHandle.left, inputHandle.right]}
        connect
      />
    </div>
  )
}
