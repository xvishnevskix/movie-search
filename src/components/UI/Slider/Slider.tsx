import { ChangeEvent, FC, useRef } from 'react';
import { SliderThumb } from './components/SliderThumb/SliderThumb';
import { IRenderThumbParams, IRenderTrackParams } from 'react-range/lib/types';
import { TextField } from '../TextField/TextField';
import { Range } from 'react-range';
import { SliderTrack } from './components/SliderTrack/SliderTrack';
import styles from './Slider.module.scss';

interface SliderProps {
	min: number;
	max: number;
	values: number[];
	step?: number;
	onChange: (values: number[]) => void;
}

export const Slider: FC<SliderProps> = ({ values, onChange, step, min, max }) => {
	const initialValueRef = useRef<number[]>(values);

	const sanitizeValues = (value: number) => {
		if (value > max) {
			return max;
		}

		return value;
	};

	const handleRenderTrack: FC<IRenderTrackParams> = ({ props, children }) => {
		return (
			<SliderTrack min={min} max={max} values={values} props={props}>
				{children}
			</SliderTrack>
		);
	};

	const handleRenderThumb: FC<IRenderThumbParams> = ({ props, value, index }) => {
		return (
			<SliderThumb
				key={index}
				props={props}
				value={value}
				initialValue={initialValueRef.current?.[index]}
			/>
		);
	};

	return (
		<div data-testid="slider" className={styles.container}>
			<div className={styles.inputs}>
				<TextField
					type="number"
					label="От"
					min={min}
					max={max}
					className={styles.input}
					value={values[0]}
					onChange={(e: ChangeEvent<HTMLInputElement>) => {
						const sanitizedValue = sanitizeValues(parseInt(e?.target.value || String(min)));
						onChange([sanitizedValue, values[1]]);
					}}
				/>
				<TextField
					type="number"
					label="До"
					min={min}
					max={max}
					className={styles.input}
					value={values[1]}
					onChange={(e: ChangeEvent<HTMLInputElement>) => {
						const sanitizedValue = sanitizeValues(parseInt(e?.target.value || String(max)));
						onChange([values[0], sanitizedValue]);
					}}
				/>
			</div>
			<Range
				step={step}
				min={min}
				max={max}
				values={values}
				onChange={onChange}
				renderThumb={handleRenderThumb}
				renderTrack={handleRenderTrack}
			/>
		</div>
	);
};
