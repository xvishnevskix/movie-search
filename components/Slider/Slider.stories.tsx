import {ComponentStory, ComponentMeta} from "@storybook/react";
import {Slider} from "./Slider";
import {useState} from "react";

export default {
    title: "Slider",
    component: Slider,
} as ComponentMeta<typeof Slider>;

const Template: ComponentStory<typeof Slider> = () => {
    const [values, setValues] = useState([0,50])

    const onChange = (values: number[]) => {
        setValues(values)
    }

    return (
        <div style={{maxWidth: "400px"}}>
            <Slider
                min={0}
                max={50}
                step={5}
                values={values}
                onChange={onChange}
            />
        </div>
    )
};

export const Example = Template.bind({});