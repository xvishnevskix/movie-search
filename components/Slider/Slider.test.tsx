import {Slider} from './Slider'
import {render, screen} from '@testing-library/react';

describe('слайдер', () => {

    const onChange = (values: number[]) => values

    test('разные тесты', () => {
        render(<Slider min={0} max={50} values={[0,50]} onChange={onChange} />)
        const element = screen.getByTestId('slider')
        expect(element).toBeInTheDocument()
        expect(element).toMatchSnapshot()
    });
})