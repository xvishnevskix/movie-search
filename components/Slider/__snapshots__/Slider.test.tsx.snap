// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`слайдер разные тесты 1`] = `
<div
  class="container"
  data-testid="slider"
>
  <div
    class="inputs"
  >
    <label
      class="label input"
    >
      <input
        class="textField"
        data-testid="input"
        value="0"
      />
    </label>
    <label
      class="label input"
    >
      <input
        class="textField"
        data-testid="input"
        value="50"
      />
    </label>
  </div>
  <div
    class="track"
  >
    <div
      aria-valuemax="50"
      aria-valuemin="0"
      aria-valuenow="0"
      class="thumb"
      draggable="false"
      role="slider"
      style="position: absolute; z-index: 0; cursor: grab; user-select: none; transform: translate(NaNpx, NaNpx);"
      tabindex="0"
    />
    <div
      aria-valuemax="50"
      aria-valuemin="0"
      aria-valuenow="50"
      class="thumb"
      draggable="false"
      role="slider"
      style="position: absolute; z-index: 1; cursor: grab; user-select: none; transform: translate(NaNpx, NaNpx);"
      tabindex="0"
    />
  </div>
</div>
`;
